tests/miral/application_selector.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "application_selector.h"
+#include "test_window_manager_tools.h"
+#include <mir/scene/session.h>
+
+using namespace miral;
+namespace mt = mir::test;
+using namespace testing;
+
+struct ApplicationSelectorTest : mt::TestWindowManagerTools
+{
+    ApplicationSelector application_selector;
+
+    ApplicationSelectorTest()
+        : mt::TestWindowManagerTools(),
+          application_selector(window_manager_tools)
+    {}
+};
+
+/// Testing if we can cycle through 3 windows and then select one
+/// with a "forward" selection strategy.
+TEST_F(ApplicationSelectorTest, run_forward)
```

Comment:
I understand why you've written it this way, but tests that check multiple state transitions are both hard to name and have complicated failure modes that can be hard to interpret.

Reply:
Agreed! I will make them more precise